pkg/gardenlet/controller/backupentry/backupentry/reconciler.go

```diff
@@ -222,12 +220,21 @@ func (r *Reconciler) reconcileBackupEntry(
 			if lastOperationState == gardencorev1beta1.LastOperationStateFailed {
 				mustReconcileExtensionBackupEntry = true
 			}
-			reconciliationSuccessful = false
 
 			lastObservedError = fmt.Errorf("extension state is not Succeeded but %v", lastOperationState)
 			if extensionBackupEntry.Status.LastError != nil {
 				lastObservedError = v1beta1helper.NewErrorWithCodes(fmt.Errorf("error during reconciliation: %s", extensionBackupEntry.Status.LastError.Description), extensionBackupEntry.Status.LastError.Codes...)
 			}
+		} else if lastOperationState == gardencorev1beta1.LastOperationStateSucceeded {
+			if updateErr := r.updateBackupEntryStatusSucceeded(ctx, backupEntry, operationType); updateErr != nil {
+				return reconcile.Result{}, fmt.Errorf("could not update status after reconciliation success: %w", updateErr)
+			}
+
+			if kutil.HasMetaDataAnnotation(&backupEntry.ObjectMeta, v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationRestore) {
+				if updateErr := removeGardenerOperationAnnotation(ctx, r.GardenClient, backupEntry); updateErr != nil {
+					return reconcile.Result{}, fmt.Errorf("could not remove %q annotation: %w", v1beta1constants.GardenerOperation, updateErr)
+				}
+			}
```

Comment:
In all cases where we set `mustReconcileExtensionBackupEntry = true`, we are not reaching this code.
And if the error occurs in the extension backupbucket in the next reconciliation, the reconciler will be invoked when it's status is updated, where we will capture this error. That's the reason why we exit reconciliation early here, https://github.com/gardener/gardener/blob/4a545ac112d135de9f3bce8a4cc66db9037de92b/pkg/gardenlet/controller/backupentry/backupentry/reconciler.go#L259-L260 too.

Reply:
Got it, thanks 🙂 

TBH, I don't like this structure too much because there is a lot of nesting esp. here where the "final step" of the reconciliation (updating `BackupBucket` to succeeded) is somewhere in between. Even though it's technically possible, readability suffers.
I think it was more understandable when this "final step" was an extra code block at the end.

Reply:
Agreed, PTAL: https://github.com/gardener/gardener/pull/7281/commits/1e3425d0c5a951d9679d964446cb40cd4b558645